cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go

```diff
@@ -339,6 +341,62 @@ func (client *autoscalingGceClientV1) DeleteInstances(migRef GceRef, instances [
 	return client.WaitForOperation(op.Name, op.OperationType, migRef.Project, migRef.Zone)
 }
 
+func (client *autoscalingGceClientV1) FetchAllInstances(project, zone, filter string) ([]GceInstance, error) {
+	registerRequest("instances", "list")
+	var instances []GceInstance
+	err := client.gceService.Instances.List(project, zone).Filter(filter).Pages(context.Background(), func(page *gce.InstanceList) error {
+		for _, gceInstance := range page.Items {
+			instance, err := externalToInternalInstance(gceInstance)
+			if err != nil {
+				klog.Errorf("Error converting instance to GceInstance: %v", err)
+				continue
+			}
+			instances = append(instances, instance)
+		}
+		return nil
+	})
+	if err != nil {
+		klog.Errorf("Failed listing Instances in zone %s: %v", zone, err)
+		return nil, err
+	}
+	return instances, nil
+}
+
+func externalToInternalInstance(gceInstance *gce.Instance) (GceInstance, error) {
+	if gceInstance == nil {
+		return GceInstance{}, fmt.Errorf("instance should not be <nil>")
+	}
+	ref, err := ParseInstanceUrlRef(gceInstance.SelfLink)
+	if err != nil {
+		return GceInstance{}, fmt.Errorf("received error while parsing of the instance url: %v", err)
+	}
+	return GceInstance{
+		Instance: cloudprovider.Instance{
+			Id: ref.ToProviderId(),
+		},
+		NumericId: gceInstance.Id,
+		Igm:       createIgmRef(gceInstance),
+	}, nil
+}
+
+func createIgmRef(gceInstance *gce.Instance) GceRef {
+	var createdBy string
+	for _, item := range gceInstance.Metadata.Items {
+		if item.Key == "created-by" && item.Value != nil {
+			createdBy = *item.Value
+		}
+	}
+	if createdBy == "" {
+		return GceRef{}
+	}
+	project, zone, name, err := ParseIgmUrl(fmt.Sprintf("https://www.googleapis.com/compute/v1/%s", createdBy))
+	if err != nil {
+		klog.Errorf("Error while parsing IGM for %v", gceInstance.SelfLink)
```

Comment:
I wonder on the log volume here. Since potentially all the instances could have a malformed igm link, should we limit logging this info, so that we don't produce too many logs in big projects?

Reply:
Added logging quota